File: 백준/레벨Gold III/10723. 판게아 1/판게아 1.ts
// @ts-ignore
import fs = require('fs');

const read = fs.readFileSync(0, 'utf-8');
const lines = read.split('\n').map((line: string) => line.split(' ').map(Number));

// 전역변수
let N: number;
let M: number;
let parent: number[];
type Edge = [number, number, number];

const insort = (arr: Edge[], value: Edge) => {
  let left = 0;
  let right = arr.length;
  while (left < right) {
    const mid = left + ((right - left) >> 1);
    if (arr[mid][0] < value[0]) left = mid + 1;
    else right = mid;
  }
  arr.splice(left, 0, value);
};

const find = (x: number): number => {
  while (x != parent[x]) {
    parent[x] = parent[parent[x]];
    x = parent[x];
  }
  return x;
};

const union = (a: number, b: number): boolean => {
  [a, b] = [find(a), find(b)];
  if (a === b) return false;

  if (a > b) parent[a] = b;
  else parent[b] = a;
  return true;
};

const kruskal = (edges: Edge[]): bigint => {
  // 부모 + 랭크 노드
  parent = Array.from({ length: N + 1 }, (_, i) => i);

  // 최소 신장 트리
  let mstCost = BigInt(0);
  let edgeCount = 0;

  for (const [w, u, v] of edges) {
    if (edgeCount >= N - 1) break;
    if (union(u, v)) {
      mstCost += BigInt(w);
      edgeCount++;
    }
  }

  return mstCost;
};

const solution = () => {
  // 1. 테스트 T회
  let lineIndex = 0;
  let [T] = lines[lineIndex++];
  while (T--) {
    [N, M] = lines[lineIndex++];

    // 2. 태초의 세계, 엣지 리스트
    const edges: Edge[] = [];
    for (let i = 1; i < N; i++) {
      const [u, c] = lines[lineIndex++];
      insort(edges, [c, i, u]); // w, u, v
    }

    // 3. 조물주가 새로이 놓은 도로
    let result = BigInt(0);
    for (let j = 1; j <= M; j++) {
      const [u, v, c] = lines[lineIndex++];
      insort(edges, [c, u, v]); // w, u, v

      // 5. XOR 연산
      result ^= kruskal(edges); // 4. 크루스칼
    }

    console.log(Number(result));
  }
};

solution();

File: 백준/레벨Gold III/10723. 판게아 1/판게아 1.py
from sys import stdin
from bisect import insort

input = stdin.readline

# 전역변수
N = None
parent = None

def find(x):
    if parent[x] < 0: return x
    parent[x] = find(parent[x])
    return parent[x]

def union(a, b):
    a, b = find(a), find(b)
    if a == b:
        return False

    if parent[a] > parent[b]:
        a, b = b, a

    parent[a] += parent[b]
    parent[b] = a
    return True

def kruskal(edges):
    global parent

    # 부모 + 랭크 노드
    parent = [-1] * (N + 1)

    # 최소 신장 트리
    mst_cost = 0
    edge_count = 0
    
    for w, u, v in edges:
        if edge_count >= N - 1: break
        if union(u, v):
            mst_cost += w
            edge_count += 1

    return mst_cost

def solution():
    global N

    # 1. 테스트 T회
    T = int(input().rstrip())
    for _ in range(T):
        N, M = map(int, input().split())

        # 2. 태초의 세계, 엣지 리스트
        edges = []
        for i in range(1, N):
            u, c = map(int, input().split())
            insort(edges, (c, i, u)) # w, u, v

        # 3. 조물주가 새로이 놓은 도로
        result = 0
        for j in range(1, M + 1):
            u, v, c = map(int, input().split())
            insort(edges, (c, u, v)) # w, u, v

            # 5. XOR 연산
            result ^= kruskal(edges) # 4. 크루스칼
            
        print(result)

solution()